### PR TITLE
Improve pathfinding performance by using a heap to store traversable polygons

### LINKED ISSUE #8496: Optimize AStar in NavAgents by using Heap for efficient min node selection
### Describe the project you are working on

This applies to any repo that uses navigation agents

### Describe the problem or limitation you are having in your project

The current implementation of the AStar algorithm in NavAgents uses a linear search to find the node with the minimum cost. This approach is inefficient, especially in complex scenes with numerous nodes, leading to performance bottlenecks during pathfinding operations.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Replace the linear search method in the AStar algorithm used by NavAgents with a heap-based approach to find the node with the minimum cost. This change will significantly improve the efficiency of the pathfinding process, reducing the computational load and enhancing the performance, particularly in scenes with a large number of polygons.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This is where the linear search takes place, the proposal here is to replace it with a `heap.pop` operation:

https://github.com/godotengine/godot/blob/923ef2b46a04e8f54c0a60bae8a0a6269fa52ed2/modules/navigation/nav_map.cpp#L358C1-L369C4

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No

### Is there a reason why this should be core and not an add-on in the asset library?

This is already an internal implementation of the Engine

### PR DESCRIPTION
**Pull request by ershn**
## Description

Improves pathfinding performance by doing 2 things.
- Makes retrieval of already visited polygons constant time instead of linear time
- Uses a heap to store the polygons to visit instead of a linked list
  - To implement this, introduces a generic heap structure that notifies of element index changes and also allows to update an arbitrary element's position (the existing functions in `SortArray` don't allow for that)

Closes https://github.com/godotengine/godot-proposals/issues/8496.

### Changes in time complexity

`n`: number of polygons visited until now (`navigation_polys`)
`m`: number of polygons marked for later visiting (`traversable_polys`)

In every scenario, `m` increases at a much slower rate than `n`.

- An individual neighbor polygon check/registration/update goes from `O(n)` to `O(log2(m))` (or `O(1)` if the polygon has already been visited and no update is needed)
- Retrieving the next polygon to visit goes from `O(m)` to `O(log2(m))`

These changes should result in a big performance improvement on big maps.

## Test project

[Navigation path finding tests.zip](https://github.com/godotengine/godot/files/13800063/Navigation.path.finding.tests.zip)

## Benchmark numbers

Below are execution times for `NavigationServer*D.map_get_path` before and after this PR on different map sizes.

Map size | Iteration count | Average time - master | Average time - PR | Speed-up
--- | --- | --- | --- | ---
15x10 cells (small) | 1000 | 0.047ms | 0.026ms | x1.80 faster
90x50 cells (normal) | 500 | 5.976ms | 0.688ms | x8.68 faster
240x160 cells (big) | 10 | 465.4ms | 11.3ms | x41.1 faster

As you can see, the bigger the map, the bigger the speed-up.

- The measurements were done on a build made with `scons platform=windows` (no debug info)
- Done on Windows 10.0.19045 with an AMD Ryzen 7 2700X (4.3 GHz max, 8 core, 16 threads)
- You can find the maps in the test project linked above

## Further improvements

There is room for further improvement here.
- Right now memory is allocated on the heap on each path request. It's rather wasteful but doesn't need to. We could keep a pool of arrays that would be reused each request.
- Finding the start and end polygon before the actual pathfinding is currently an `O(n)` operation. We will invariably loop over every polygon each request. We could use a specialized data structure to improve performance here.
- With the current API, even if we're interested only in paths that actually reach the target, a path to the closest reachable polygon will always be calculated if the target is unreachable. That's two pathfinding requests even if we only want one. We could add a new option/method to cover this use case. That one requires a proposal though.

I intend to work on these if there is no problem with this PR.

### CODE REVIEW

File: core/templates/heap.h
Comment (AThousandShips) [inline core/templates/heap.h:58]: Should have an `is_empty` like the other types

File: core/templates/heap.h
Comment (AThousandShips) [inline core/templates/heap.h:87]:
```suggestion
		if (!_shift_up(p_index)) {
			_shift_down(p_index);
		}
```
Or something similar, the original syntax has very poor readability IMO

**Comment by ershn** [comment]
I realized that the current implementation doesn't take care of multithreading.
Trying to find a path from multiple threads on the same map at the same time will currently cause a crash.

I'm turning this PR back into a draft until I fix the problem.

**Comment by ershn** [comment]
I changed the approach to be multithreading friendly. Pathfinding requests now leave `Polygon` objects untouched.

1. On map construction (`NavMap::sync`), assign each `Polygon` an id which is its index in the vector
2. On entering `map_get_path`, create a `NavigationPoly` vector of the same size as the one above
3. Use the `Polygon` id to find the corresponding `NavigationPoly` in the vector and vice versa

Since it's not possible anymore for the `NavigationPoly` vector to grow in size during pathfinding, I also changed it so that I now store pointers in the heap instead of vector indexes. It improved readability and also slightly increased performance.

I added a multithreading test to the test project and also rebased on master (checked that https://github.com/godotengine/godot/issues/85238 is still fixed).

**Comment by smix8** [comment]
I suggest moving this to a navigation module internal class/struct instead of a core template.

I talked with some core devs and there is a preference for a more "local" solution that does not touch core.

**Comment by smix8** [comment]
With the merge window for 4.4 now open I would like to get this in but it requires rebase and the mentioned move of the new heap class from core to the navigation module. I would suggest to move it to the nav_utils.h file in the gd namespace where everything else currently is that is used across the navsystem.

If you are still around @ershn and want to do these changes for the PR appreciated, else just let me know and I will salvage the PR and give you credit. Either way thank you for doing the main work on this optimization and all the performance testing.

**Comment by ershn** [comment]
Sorry for the delay. I moved the heap class to `nav_utils.h`, rebased on master, squashed the commits and checked that nothing broke after that.
I wasn't entirely sure where to put the new unit test file so feel free to change that if necessary.

Thank you for the review smix8 !

**Comment by smix8** [comment]
> I wasn't entirely sure where to put the new unit test file so feel free to change that if necessary.

@Scony is usually the test maniac ... ;)
... but I would say those tests would belong to the `tests/servers/test_navigation_server_3d.h` file.

**Comment by ershn** [comment]
Thank you for the tests !

The bug probably happened because the synthetic polygons generated for navigation links were not being taken care of during map sync. Their id was never updated from the default value of UINT32_MAX which resulted in invalid array accesses when trying to process them during pathfinding.

This should be fixed now. Could you test again with your test scenes ?

> but you can imo also just leave it in that separated file as is for now
Ok, I'll leave the tests there for the time being then.

**Comment by smix8** [comment]
> This should be fixed now. Could you test again with your test scenes ?

@ershn  Yeah that was it, no more crashes or errors (so far 😝)

**Review by smix8 — APPROVED** [review]

File: modules/navigation/nav_utils.h
Comment (ershn) [reply modules/navigation/nav_utils.h]: Fixed.

File: modules/navigation/nav_utils.h
Comment (ershn) [reply modules/navigation/nav_utils.h]: Fixed.

File: modules/navigation/nav_utils.h
Comment (ershn) [reply modules/navigation/nav_utils.h]: Fixed.

File: tests/servers/test_navigation_heap.h
Comment (ershn) [reply tests/servers/test_navigation_heap.h:1]: Ok. I moved it into `tests/servers/test_navigation_server_3d.h` like smix8 suggested if that's what you meant.

Comment (AThousandShips) [reply tests/servers/test_navigation_heap.h:1]: I meant it should be moved to `modules/navigation/tests`, but I think that's okay too

**Review by Scony — COMMENTED** [review]
In my view, the heap implementation is generic enough so that should placed be within `core/templates`. @smix8 who are the "core devs" that suggested otherwise?

If not `core/templates` then I don's see a reason why this fairly-generic class wouldn't be placed in some separate file like `modules/navigation/heap.h` or so. Same goes for tests - `test_navigation_server_3d.h` which is actually a bunch of **module tests** is not a good place for heap **unit tests**. I'd at least extract it to `tests/servers/test_navigation_server_3d_heap.h` and include in `tests/servers/test_navigation-server_3d.h`.

**Comment by smix8** [comment]
> In my view, the heap implementation is generic enough so that should placed be within `core/templates`. @smix8 who are the "core devs" that suggested otherwise?

That was when I asked reduz with 1-2 other core people around for a review on this PR months ago. They all said this should be local to the navigation module and not part of core which is why I made that comment back in April to unblock this PR.

**Comment by Scony** [comment]
Ok, let's do the refactoring in the followup PR.

**Review by Scony — APPROVED** [review]

**Comment by kleonc** [comment]
> In my view, the heap implementation is generic enough so that should placed be within `core/templates`.

@Scony See [this](https://docs.godotengine.org/en/4.3/contributing/development/best_practices_for_engine_contributors.html#prefer-local-solutions). Currently it's used only by the navigation so adding to core is not justified. If something else will need it as well then it will probably be fine to move it to core.
(I'm not a core team member but IIRC that's the reasoning.)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
